doc/api/async_hooks.md

```diff
@@ -579,6 +579,105 @@ const server = net.createServer((conn) => {
 Promise contexts may not get valid `triggerAsyncId`s by default. See
 the section on [promise execution tracking][].
 
+### Class: `AsyncLocal`
+
+<!-- YAML
+added: REPLACEME
+-->
+
+This class can be used to store a value which follows asynchronous execution
+flow. Any value set on an `AsyncLocal` instance is propagated to any callback
+or promise executed within the flow. Because of that, a continuation local
+storage can be build with an `AsyncLocal` instance. This API is similar to
+thread local storage in other runtimes and languages.
+
+The implementation relies on async hooks to follow the execution flow.
+So, if an application or a library does not play nicely with async hooks,
+the same problems will be seen with the `AsyncLocal` API. In order to fix
+such issues the `AsyncResource` API should be used.
+
+The following example shows how to use `AsyncLocal` to build a simple logger
+that assignes ids to HTTP requests and includes them into messages logged
+within each request.
+
+```js
+const http = require('http');
+const { AsyncLocal } = require('async_hooks');
+const asyncLocal = new AsyncLocal();
+function print(msg) {
+  const id = asyncLocal.unwrap();
+  console.log(`${id !== undefined ? id : '-'}:`, msg);
+}
+let idSeq = 0;
+http.createServer((req, res) => {
+  asyncLocal.store(idSeq++);
+  print('start');
+  setImmediate(() => {
+    print('finish');
+    res.end();
+  });
+}).listen(8080);
+http.get('http://localhost:8080');
+http.get('http://localhost:8080');
+// Prints:
+//   0: start
+//   1: start
+//   0: finish
+//   1: finish
+```
+
+#### `new AsyncLocal()`
```

Comment:
I am curious how this relates to the domain problem as it does not enforce scoping. Is that a non-goal here? 
In that case, it might be clearer if the doc highlights the pitfalls of the absence of scoping (which is an important feature for such API IMO).

Reply:
The "scoping" is intentionally similar to lexical scoping--anywhere with access to the `asyncLocal` instance has access to the value it contains. At any given point in the lifetime of the `asyncLocal`, the value can be _branched_ by storing a new value, resulting in all following async tasks created in the remainder of that sync tick to follow from that branch. This is _similar_ to the domain problem, but not quite the same as domains suffer from multiple-entry problems as there is logic that can be specifically _depended_ on in enter/exit points, whereas here the only thing that can be "depended" on is the availability of the object, which is intentionally accessible in this way.

The exit "cleanup" that most CLS implementations tend to do is something like `store[key] = undefined` to "restore" to a blank state after entering does something like `store[key] = getParentContext()`. However there's actually _never_ any point at which the `undefined` state would be reachable because it would _always_ be in an "entered" state for the remainder of the async execution tree unless disabled which typically needs to be cleaned up differently anyway. Because there's now a _resource_ stack, we can actually just use the resource itself to represent the "entered" or "exited" state and never actually need an _event_ for it, relying only on an eager propagation in the "init" event and then a simple lookup later. The resource will remain constant for the entire remainder of the callback tick, so the descending tree is stable, it's only the sync tick where a branch/transition occurs that could be considered "unstable", but it functions that way intentionally. By branching mid-way through the tick, it's exactly the same as a `run(callback)` method, except it prohibits the possible edge case of a `run(...)` with more code after it that might be _expected_ to be in-scope, but is not. Consider an express middleware:

```js
const cls = new SomeCLS()

app.use((req, res, next) => {
  cls.run(store => {
    store.set('requestId', req.id)
  })
  next()
})

app.get('/', (req, res) => {
  const store = cls.getStore()
  res.end(store.get('requestId'))
})
```

This will not work as expected, because the `next()` call was made _outside_ of the run. I see user errors like this _all the time_, and they can be eliminated by changing from an enter/exit machine to a simpler transition state--you're never "outside", just transitioned to a different level. With the state transition model, you don't _need_ an explicit `run(...)` call as it _automatically_ ensures that any _logically following_ code descends from the branch made at the point when the local value was stored.

If you think about it, the "before" and "after" pairing in async_hooks is actually _kind of_ redundant as you can be certain that _any_ time a "before" is triggered it implies the "after" state of the previous callback tick has already been reached. Of course there are ergonomics reasons for this--using any of the id/resource fetching APIs would not give you the data you need to handle cleanup normally done in "after" events if you tried to directly move that logic to the "before" events. Also, _technically_ you can nest enter/exits syncronously with `AsyncResource`, but that's an edge case that is already worked around automatically through state transitions gained by `executionAsyncResource` observing the resource stack.

Reply:
Fully agree here.

> If you think about it, the "before" and "after" pairing in async_hooks is actually _kind of_ redundant...

I tried to model the transitions in https://github.com/nodejs/node/pull/27172 (similar as .NET but they even have an ExecutionContext class to signal) to hide the before/after callbacks of async hooks but it didn't make much people happy.